aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,645 @@
+          return pools_ptr[x] == pools_ptr[y];
+        });
+    auto new_sz = thrust::distance(count_set.begin(), new_end.second);
+    count_set.resize(new_sz);
+  }
+  thrust::device_vector<int64_t> device_pool_offsets = count_set;
+
+  thrust::exclusive_scan(
+      device_pool_offsets.begin(),
+      device_pool_offsets.end(),
+      device_pool_offsets.begin());
+
+  return std::make_tuple(
+      sorted_indices,
+      device_pool_offsets,
+      count_set,
+      mx_rows,
+      log_softmax_max_buffer);
+}
+
+template <typename scalar_t, bool LogSoftMax>
+void cuda_sparse_coo_softmax(
+    Tensor& output,
+    const Tensor& input,
+    const int64_t dim) {
+  auto sparse_dim = input.sparse_dim();
+  auto indices = input._indices().contiguous();
+  auto values = input._values().contiguous();
+  auto out_values = output._values();
+  auto out_indices = output._indices();
+  out_values.resize_as_(values);
+  out_indices.resize_as_(indices);
+  out_indices.copy_(indices);
+
+  if (dim >= sparse_dim) {
+    if (LogSoftMax) {
+      auto new_values = log_softmax_cuda(values, dim - sparse_dim + 1, false);
+      out_values.copy_(new_values);
+    } else {
+      auto new_values = softmax_cuda(values, dim - sparse_dim + 1, false);
+      out_values.copy_(new_values);
+    }
+    return;
+  }
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  auto nnz = values.size(0);
+  auto sizes = input.sizes();
+  auto nvalues = get_nvalues(sizes, sparse_dim);
+
+  /* Prepare accessors */
+  auto values_2 = values.view({nnz, nvalues});
+  auto values_accessor = values_2.packed_accessor<scalar_t, 2>();
+
+  auto out_values_2 = out_values.view({nnz, nvalues});
+  auto out_values_accessor = out_values_2.packed_accessor<scalar_t, 2>();
+
+  thrust::device_vector<int64_t> sorted_indices;
+  thrust::device_vector<int64_t> device_pool_offsets;
+  thrust::device_vector<int64_t> device_pool_sizes;
+  thrust::device_vector<scalar_t*> mx_rows;
+  thrust::device_vector<scalar_t> log_softmax_max_buffer;
+
+  std::tie(
+      sorted_indices,
+      device_pool_offsets,
+      device_pool_sizes,
+      mx_rows,
+      log_softmax_max_buffer) =
+      compute_pool_max<scalar_t, true, LogSoftMax>(
+          indices, values_2, sizes, nvalues, dim);
+
+  auto pool_size = device_pool_offsets.size();
+  thrust::device_vector<scalar_t> exp_sums_rows(nvalues * pool_size, 0);
+  int block_size = getNumThreads(pool_size);
+  const int grid_size = (pool_size + block_size - 1) / block_size;
+
+  cuda_sparse_coo_softmax_kernel<scalar_t, LogSoftMax>
+      <<<grid_size, block_size, 0, stream>>>(
+          sorted_indices.data().get(),
+          pool_size,
+          device_pool_sizes.data().get(),
+          device_pool_offsets.data().get(),
+          nvalues,
+          mx_rows.data().get(),
```

Comment:
Shouldn't `log_softmax_max_buffer` be used here instead of `mx_rows`?

Reply:
As `mx_rows` is a valid index reference to `log_softmax_max_buffer`, no changes is required here.